[Roles] Added transformation of application * privilege to all

elena-shostak opened the pull request:
## Summary

Added transformation of application wildcard `*` privilege to `all` to correctly filter and display roles as `superuser`.

<details>
  <summary>[Screenshot] Kibana superuser role before change</summary>
 
<img width="1508" alt="Screenshot 2024-04-23 at 11 20 31" src="https://github.com/elastic/kibana/assets/165678770/594cf27c-64a7-49cf-bf4b-c63adca76a55">

</details>

<details>
  <summary>[Screenshot] Kibana superuser role after change</summary>
<img width="1504" alt="Screenshot 2024-04-23 at 11 17 54" src="https://github.com/elastic/kibana/assets/165678770/96dbdaa6-f5c9-4644-82ca-f88a0f6a15b3">
</details>

### Checklist

- [X] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios


### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

__Fixes: https://github.com/elastic/kibana/issues/106561__

## Release note
Added transformation of application wildcard `*` privilege to `all` to correctly filter and display roles as `superuser`.

elena-shostak committed:
[Roles] Added transformation of application * privilege to all

elasticmachine commented:
Pinging @elastic/kibana-security (Team:Security)

jeramysoucy reviewed (changes requested):
Nice work, Elena!

I just found one issue. If I create a role via ES API with wildcard application privileges (like `superuser` and `system_indices_superuser` both have), and then load and save the role in the UI (making no changes), the role ends up with an additional entry that correlates to the translated 'all' privilege.

>     "applications": [
>         {
>             "application": "*",
>             "privileges": [
>                 "*"
>             ],
>             "resources": [
>                 "*"
>             ]
>         }
>     ],

becomes 

>     "applications": [
>         {
>             "application": "kibana-.kibana",
>             "privileges": [
>                 "all"
>             ],
>             "resources": [
>                 "*"
>             ]
>         },
>         {
>             "application": "*",
>             "privileges": [
>                 "*"
>             ],
>             "resources": [
>                 "*"
>             ]
>         }
>     ],

And it also shows up in the UI:
![Screenshot 2024-04-24 at 11 21 33 AM](https://github.com/elastic/kibana/assets/103939324/7a667549-5936-4fcd-bef9-166c4cb4043e)

It will also occur if a user edits the translated privileges in the UI. When the privilege is updated, it will retain the original wildcard as well as a new 'kibana':'all'.

We may want to intercept this on save/update and have a special case for when a role already contains the wild card so it is either retained (if assigning all) or overwritten (if changed), and potentially display a message to the user when a role contains the superuser style wildcard.

cc: @legrego any thoughts?

legrego commented:
> If I create a role via ES API with wildcard application privileges (like superuser and system_indices_superuser both have), and then load and save the role in the UI (making no changes), the role ends up with an additional entry that correlates to the translated 'all' privilege.

Great find, @jeramysoucy. Thanks for your attention to detail here.


> We may want to intercept this on save/update and have a special case for when a role already contains the wild card so it is either retained (if assigning all) or overwritten (if changed), and potentially display a message to the user when a role contains the superuser style wildcard.
> 
> cc: @legrego any thoughts?

I'm not opposed to exploring some sort of interception if it's not an unreasonable amount of effort, and not too hacky. Otherwise, we should consider alternate approaches.

elena-shostak committed:
[Roles] Moved wildcard normalization to UI

elena-shostak committed:
Disallow editing wildcard app privilege

elena-shostak commented:
@jeramysoucy Thanks for bringing it in. Looked carefully at the setup, it seems that my normalization of application wildcard privilege on the server brought more problems than it eventually solved.

Discussed it with @legrego, we've decided to not normalize it  to `all`, but rather display as `*` and disallow edit of such privileges in the UI. See attached screenshot below.

<details>
  <summary>[Screenshot] Kibana wildcard role display</summary>
 
<img width="1503" alt="Screenshot 2024-04-24 at 15 54 53" src="https://github.com/elastic/kibana/assets/165678770/63575a57-6c8a-43e1-91ce-0caebd2d3e9f">


</details>

jeramysoucy commented:
> Discussed it with @legrego, we've decided to not normalize it to all, but rather display as * and disallow edit of such privileges in the UI. See attached screenshot below.

Sounds good. Do we want to disallow the delete action as well?

elena-shostak committed:
[Roles] Disabled delete action

elena-shostak committed:
[Roles] Added test

elena-shostak committed:
fixup

jeramysoucy reviewed (changes requested):
Sorry in advance. I found a couple of things.

- Can we notify the user somehow that a base wildcard privilege is special? The action icons are removed, but there is nothing else to let the user know what's going on. Maybe an icon tip where the action icons usually are?  cc: @legrego do we need to check with the design team or doc team for approach/wording?
- If I view a role with the base wildcard and click Update, there is an error:
![Screenshot 2024-04-25 at 1 49 43 PM](https://github.com/elastic/kibana/assets/103939324/18da8ce9-53e9-4a22-aeab-674eda69867d)
This is tricky, because if we disable the update button then no changes can be made to ES privileges for the role. So I think we have to trap the Kibana base wildcard on update, and only apply any ES changes.
- If I view a role with the base wildcard, I still have options to add a Kibana privilege and view a privilege summary.
![Screenshot 2024-04-25 at 2 02 54 PM](https://github.com/elastic/kibana/assets/103939324/5d2132a5-5b50-4759-b6fc-c035c3343c05)
The summary shows all "none" (expected since this is a special case), but maybe we should remove the ability to view this is the base wildcard is the only privilege?
![Screenshot 2024-04-25 at 2 03 03 PM](https://github.com/elastic/kibana/assets/103939324/361ef8aa-d3dd-44a4-a3a9-6a87d44d85bd)
I think both of these options, `Add Kibana privilege` and `View privilege summary`, should be removed for roles with the base wildcard.

jeramysoucy commented on x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/privilege_form_calculator/privilege_form_calculator.ts:
```diff
@@ -27,9 +27,21 @@ export class PrivilegeFormCalculator {
+   *
+   * @param privilegeIndex the index of the kibana privileges role component
+   */
+  public isWildcardBasePrivilege(privilegeIndex: number) {
```
Nit: missing from unit tests

elena-shostak committed:
[Roles] Disallowed edit actions for role with wildcard privilege

elena-shostak committed:
[Roles] Added test for PrivilegeFormCalculator

elena-shostak commented:
@jeramysoucy @legrego 

Changes addressed:

- [X] Added tests to cover scenarios with edit/delete icon actions
- [X] Added tests to cover scenarios with `View privilege summary`, `Add Kibana privilege` actions
- [X] Made the role non editable if there is a base privilege with the wildcard present.

See screenshot
<img width="1508" alt="Screenshot 2024-04-25 at 19 07 21" src="https://github.com/elastic/kibana/assets/165678770/8783b2a7-bf23-4390-89c6-80ab0862623f">

Let me know if that works and makes sense.

elena-shostak committed:
[Roles] fixup

jeramysoucy commented on x-pack/plugins/security/public/management/roles/edit_role/edit_role_page.test.tsx:
```diff
@@ -744,6 +744,29 @@ describe('<EditRolePage />', () => {
+    await waitForRender(wrapper);
+
+    expect(wrapper.find('[data-test-subj="privilegeEditAction-0"]')).toHaveLength(0);
+    expect(wrapper.find('[data-test-subj="privilegeDeleteAction-0"]')).toHaveLength(0);
+  });
```
Can we also add a check that this is rendered as read-only?

```suggestion
    expect(wrapper.find('[data-test-subj="privilegeDeleteAction-0"]')).toHaveLength(0);
    expectReadOnlyFormButtons(wrapper);
  });
```

elena-shostak replied:
Sure, updated, thanks!

jeramysoucy reviewed (approved):
I just had one suggestion in the new test, to check for a read-only button render. I am approving now so that i don't hold up this PR any further.

I think making roles with the app wildcard privilege read-only in the UI is fine. These roles could only have been create/updated via the ES API anyway. I'm not sure how I feel about there not being any message to clarify this to the end user, but this is something we can come back to if needed.

Great work, Elena!

elena-shostak committed:
Updated test

elena-shostak commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into 106561-superuser-role

kibanamachine committed:
Merge branch 'main' into 106561-superuser-role

elena-shostak commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into 106561-superuser-role

elena-shostak committed:
Type fix

jasonrhodes commented:
Thank you all so much for this change, I'd been following an earlier issue that this PR closed and it was very happy news to see! <3
